Change generate-copyright to generate HTML, with cargo dependencies included

jonathanpallant opened the pull request:
`x.py run generate-copyright` now produces `build/COPYRIGHT.html`. This includes a new format for in-tree dependencies, and also adds out-of-tree cargo dependencies.

After consulting expert opinion, I have elected to include every top-level:

* `*NOTICE*`
* `*AUTHOR*`
* `*LICENSE*`
* `*LICENCE*`, and
* `*COPYRIGHT*` file I can find - case-insensitive.

This is because the cargo package metadata's `author` field is not a list of copyright holders and does not meet the requirements of the Apache-2.0 license (which says you must include a NOTICE file with the binary if one was supplied by the author) nor the MIT license (which says you must include 'the above copyright notice').

I believe it would be appropriate to include this file with every Rust release, in order to do an even better job of appropriately recognising the efforts of the authors of the first-party and third-party libraries we are using here.

The output includes something like 524 copies of the Apache-2.0 text because they are not all identical. I think I count about 50 different variations by shasum - some differ in whitespace, while some have the boilerplate block at the bottom erroneously modified (don't modify the copy in the license, modify the copy you paste into your own source code!). Running `gzip` on the HTML file largely makes this problem go away, and the average browser is far happier with a ~6 MiB HTML file than the average Markdown viewer is with a ~6 MiB markdown file. But, if someone wants to, do they could submit a follow-up which de-dups the license text files and adds back-links to earlier identical copies (for some value of 'identical copy').

```console
$ xpy run generate-copyright
$ cd build
$ gzip -c COPYRIGHT.html > COPYRIGHT.gz
$ xz -c COPYRIGHT.html > COPYRIGHT.xz
$ ls -lh COPYRIGHT.*
-rw-r--r--  1 jonathan  staff   241K 29 Jul 17:19 COPYRIGHT.gz
-rw-r--r--@ 1 jonathan  staff   6.6M 29 Jul 11:30 COPYRIGHT.html
-rw-r--r--  1 jonathan  staff    59K 29 Jul 17:19 COPYRIGHT.xz
```

Here's an example [COPYRIGHT.gz](https://github.com/user-attachments/files/16416147/COPYRIGHT.gz).

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r​? <reviewer name>
-->
<!-- homu-ignore:end -->

jonathanpallant committed:
Update generate-copyright

This tool now scans for cargo dependencies and includes any important looking license files.

We do this because cargo package metadata is not sufficient - the Apache-2.0 license says you have to include any NOTICE file, for example. And authors != copyright holders (cargo has the former, we must include the latter).

jonathanpallant committed:
generate-copyright: Produce HTML, not Markdown

This format works better with large amounts of structured data.

We also mark which deps are in the stdlib

Kobzol commented on src/tools/generate-copyright/src/cargo_metadata.rs:
```diff
@@ -0,0 +1,196 @@
+
+/// Describes one of our dependencies
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Dependency {
```
We already have https://crates.io/crates/cargo_metadata in our dependency tree. Maybe we could save some custom code by using it? It can extract metadata from a Cargo package, and it defines a bunch of useful Cargo types.

jonathanpallant replied:
The types weren't useful, but the function really was. Thanks!

Kobzol commented on src/tools/generate-copyright/src/main.rs:
```diff
@@ -1,54 +1,114 @@
+///
+/// You should probably let `bootstrap` execute this program instead of running it directly.
+///
+/// Run `x.py run generate-metadata`
```
`generate-metadata` -> `generate-copyright`

Kobzol commented on src/tools/generate-copyright/src/main.rs:
```diff
@@ -1,54 +1,114 @@
+
+    let root_path = std::path::absolute(".")?;
+    let workspace_paths = [
+        Path::new("./Cargo.toml"),
```
I think that we have more workspaces. E.g. recently `src/tools/rustbook` was added as a separate workspace. We should probably somehow pass their list from bootstrap, to avoid forgetting about them? :thinking:

pietroalbini replied:
I would love to see a central slice in bootstrap listing all workspaces :D

Kobzol replied:
Hmm, I took a look and "all workspaces" is not as easy as it sounds :laughing: We currenty have the main workspace, bootstrap workspace, `rustbook` workspace, but then also a bunch of workspaces for individual tools, such as RA. I'm not sure if we can generalize that list for all use-cases. For example, this code probably doesn't want to include the workspace of bootstrap and RA (?).

jonathanpallant replied:
I think this tool should include anything we 'ship' with each 'Rust Release', for some value of 'ship' and some value of 'Rust Release'.

bjorn3 replied:
https://github.com/rust-lang/rust/pull/128588 dedups the list of workspaces to vendor between the two locations where we run `cargo vendor`. Maybe `generate-copyright` should use the same list of workspaces. That is the set of workspaces that may actually be used by `./x.py dist` and `./x.py test`. Everything else would fail to build on offline systems. For example `library/stdarch` workspace is not meant to be used. Some of it's members are used as dependencies of the standard library, but the rest aren't vendored and bootstrap doesn't expose a way to run the stdarch tests either. Tidy also currently has a list of workspaces for which it checks licenses. That one should probably also be deduped with `cargo vendor` and `generate-copyright`.

thejpster replied:
That sounds reasonable but I'm not sure I know enough about the internal org of this repo to do it. Also I'm on vacation now.

@pietroalbini ?

Kobzol commented on src/tools/generate-copyright/src/main.rs:
```diff
@@ -71,93 +84,100 @@ fn main() -> Result<(), Error> {
-fn render_deps<'a, 'b>(
-    deps: impl Iterator<Item = &'a cargo_metadata::Dependency>,
-    buffer: &'b mut Vec<u8>,
+fn render_deps(
```
This generates some weird looking HTML, both in terms of tags:
![image](https://github.com/user-attachments/assets/9d880ed5-9684-4dea-92fd-75f9b70e077c)
and content (is the strikethrough expected?):
![image](https://github.com/user-attachments/assets/95860013-4f2a-49c7-b05a-59912b8366ae)

jonathanpallant replied:
Ah, that's a less than in a field not being escaped and the first letter of the email being picked up as a tag.

Kobzol replied:
This specific case/bug could be fixed, of course, but I think in general it would be both safer and easier to just use something automatic to generate the HTML than to code this from scratch, as I noted in my other comment.

Kobzol reviewed (commented):
Hi, thanks for the PR! Left some comments and nits.

From a high-level point of view, I have some doubts about the second commit. I don't like the approach of building a HTML manually out of strings very much. It seems quite simple to produce mangled/invalid HTML content or forget to escape something (especially since we're including data from Cargo manifests of external packages!).

Could we create something that is more structured and then generate a HTML page out of it automatically? We could either just generate the HTML from a Markdown file or generate a bunch of Rust structs and then use some simple Jinja/Rinja/Tera/Askama/whatever template to actually build the website, with proper escaping and less custom code. We already have a bunch of useful third-party crates in our dependency tree, so I don't think that it's needed to create our own manual code for rendering a HTML page.

jonathanpallant committed:
generate-copyright: Fix typo

jonathanpallant committed:
generate-copyright: use cargo-metadata

jonathanpallant commented:
I'll review what's in the tree already and see if I can pick something to do the HTML escaping for sure, and perhaps the rendering too.

Kobzol commented:
We already have https://github.com/rinja-rs/rinja for template rendering or https://github.com/pulldown-cmark/pulldown-cmark for converting Markdown into HTML.

jonathanpallant commented:
Markdown might be a little tricky because some of the license texts are almost, but not quite, valid Markdown. And I already basically have what I want in Rust structs - so I'll look at rinja. Thanks!

jonathanpallant committed:
generate-copyright: use rinja to format the output

I can't find a way to derive rinja::Template for Node - I think because it is a recursive type. So I rendered it manually using html_escape.

jonathanpallant committed:
REUSE.toml: Copyright text isn't parsed as Markdown.

jonathanpallant commented:
OK, I tried to use rinja but I couldn't make it render a recursive type. So I manually rendered the `Node` type, and used `html_escape` to escape the output.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10167955852/job/28121456092) [(plain)](https://github.com/rust-lang/rust/commit/041e4ba2fa3438395ee98561f88eb5a86fbfc420/checks/28121456092/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.854 Building wheels for collected packages: reuse
#16 2.855   Building wheel for reuse (pyproject.toml): started
#16 3.101   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.102   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.103   Stored in directory: /tmp/pip-ephem-wheel-cache-ilg9u5u5/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.105 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.499 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.499 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.6s
```

</details>

jonathanpallant committed:
generate-copyright: Render Node with rinja too.

jonathanpallant commented:
OK, I made it work! No idea what I was doing wrong last time - the errors you get when a template fails to compile are inscrutable.

thejpster commented on src/tools/generate-copyright/src/cargo_metadata.rs:
```diff
@@ -0,0 +1,175 @@
+            }
+            if keep {
+                if metadata.is_dir() {
+                    // scoop up whole directory
```
Oops. I should scoop up the whole directory here.

thejpster commented on src/tools/generate-copyright/Cargo.toml:
```diff
@@ -2,10 +2,15 @@
+rinja = "0.2.0"
 serde = { version = "1.0.147", features = ["derive"] }
 serde_json = "1.0.85"
+tempfile = "3"
```
I don't think I actually use this one any more.

bors commented:
:umbrella: The latest upstream changes (presumably #128634) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jonathanpallant committed:
generate-copyright: gather files inside interesting folders

jonathanpallant committed:
Update to rinja 0.3

jonathanpallant commented:
I've rebased on main locally, and this tool is now broken:

```console
$ xpy run generate-copyright
Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.18s
Building stage0 tool collect-license-metadata (aarch64-apple-darwin)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.21s
gathering license information from REUSE
finished gathering the license information from REUSE in 33.71s
Building stage0 tool generate-copyright (aarch64-apple-darwin)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.22s
Vendoring deps into /Users/jonathan/Documents/open-source/rust/build/vendor...
error: current package believes it's in a workspace when it's not:
current:   /Users/jonathan/Documents/open-source/rust/./library/std/Cargo.toml
workspace: /Users/jonathan/Documents/open-source/rust/Cargo.toml

this may be fixable by adding `library/std` to the `workspace.members` array of the manifest located at: /Users/jonathan/Documents/open-source/rust/Cargo.toml
Alternatively, to keep it out of the workspace, add the package to the `workspace.exclude` array, or add an empty `[workspace]` table to the package's manifest.
Error: Failed to complete cargo vendor

Command CARGO="/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage0/bin/cargo" DEST="/Users/jonathan/Documents/open-source/rust/build/COPYRIGHT.html" DYLD_LIBRARY_PATH="/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage0-bootstrap-tools/aarch64-apple-darwin/release/deps:/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage0/lib" LICENSE_METADATA="/Users/jonathan/Documents/open-source/rust/build/license-metadata.json" OUT_DIR="/Users/jonathan/Documents/open-source/rust/build" RUSTC="/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage0/bin/rustc" "/Users/jonathan/Documents/open-source/rust/build/aarch64-apple-darwin/stage0-tools-bin/generate-copyright" (failure_mode=Exit) did not execute successfully.
Expected success, got exit status: 1
Created at: src/core/build_steps/tool.rs:1113:23
Executed at: src/core/build_steps/run.rs:222:13

Build completed unsuccessfully in 0:00:39
```

I'm not sure what I've done wrong, and I'm not out on vacation for three weeks.

jonathanpallant committed:
Apparently library/std is now part of a workspace at library/

jonathanpallant commented:
Ah, @Veykril helped me fix it. Apparently libstd joined a workspace at `./library/Cargo.toml`

Kobzol reviewed (commented):
Yes, this happened in https://github.com/rust-lang/rust/pull/128534.

I wonder if this is expected? The root level files should also be under MIT/Apache-2.0.

![image](https://github.com/user-attachments/assets/67636083-46b8-4726-a039-e74124f517ec)

thejpster commented on src/tools/generate-copyright/src/cargo_metadata.rs:
```diff
@@ -150,24 +152,38 @@ fn load_important_files(
+                        let inner_filename_str = inner_filename.to_string_lossy();
+                        let qualified_filename =
+                            format!("{}/{}", lc_filename_str, inner_filename_str);
+                        println!("Scraping {}", qualified_filename);
```
I left them in because I thought it was useful to know which files were being picked up and packaged. Happy to take them out though.

Kobzol replied:
Ok, then let's leave it as it is :)

jonathanpallant commented:
What does your license-metadata.json say?

Mine says:

```text
 205   │         "license": {
 206   │           "copyright": [
 207   │             "The Rust Project Developers (see https://thanks.rust-lang.org)"
 208   │           ],
 209   │           "spdx": "Apache-2.0 OR MIT"
 210   │         },
 211   │         "name": ".",
 212   │         "type": "directory"
 ```

My HTML looks like this:

<img width="571" alt="image" src="https://github.com/user-attachments/assets/f810f88f-70d0-4b13-adcc-4499d34d6e18">

Kobzol commented:
```json
{
  "files": {
    "children": [
      {
        "children": [...],
        "license": {
          "copyright": [
            "NONE"
          ],
          "spdx": "NONE"
        },
        "name": ".",
        "type": "directory"
      }
    ],
    "type": "root"
  }
}
```
Interesting :thinking:

thejpster commented:
Ok, not generate-copyright's problem then. 

The license data generator does require reuse-tool 4.0, otherwise it'll ignore the new REUSE.toml file.

Kobzol commented:
Hmm, I have `reuse 4.0.3` (the Python tool). I'll try to debug what might be causing this.

Kobzol commented:
I think it's some missing logic in collect-license-metadata that doesn't count with the fact that I have untracked files in the root. I suppose that this doesn't need to block this PR. Do you want to make some modifications here? Otherwise I think we can ship this.

thejpster commented:
I have no further changes to generate-copyright.

Kobzol commented:
Ok then, thank you!

@bors r+

bors commented:
:pushpin: Commit 99579f3ec1ee070d3d8203df1168f948b05b48dc has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 99579f3ec1ee070d3d8203df1168f948b05b48dc -->
<!-- homu: {"type":"Approved","sha":"99579f3ec1ee070d3d8203df1168f948b05b48dc","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bjorn3 commented:
Did you see https://github.com/rust-lang/rust/pull/128353#discussion_r1707459067?

Kobzol commented:
Ah, right. Yeah, we now have a bunch of "workspace lists", deduping everything would be nice, although I'm not sure if we really use the same set of workspace in *all* places. In any case, I don't think that this needs to block this PR. I'll try to take a look at it.

bjorn3 commented:
Both tidy and generate-copyright should probably use the exact same set to ensure generate-copyright will never show a license forbidden by tidy. `cargo vendor` should be a (non-strict, aka both sides of the subset operator may be identical) subset of what tidy and generate-copyright use. If we vendor anything not known by generate-copyright, the latter would miss some licenses included in the source tarball.

Kobzol commented:
> Both tidy and generate-copyright should probably use the exact same set to ensure generate-copyright will never show a license forbidden by tidy. `cargo vendor` should be a (non-strict, aka both sides of the subset operator may be identical) subset of what tidy and generate-copyright use. If we vendor anything not known by generate-copyright, the latter would miss some licenses that would be included in the source tarball.

This was implemented in https://github.com/rust-lang/rust/pull/137020.
